### ISSUE TITLE
Building Istio images based on gcr.io/distroless/*

### ISSUE THREAD

**Issue by loewenstein** [istio/istio#12122]
**Describe the feature request**
In order to use Istio in production, we need to scan and assess Istio images for vulnerabilities.
We would have an easier job, if Istio images would be based on smaller base images or preferably no base image at all.

**Describe alternatives you’ve considered**
We could accomplish this also by stripping down the existing images on our own, to remove executables with known vulnerabilities.

**Additional context**
Vulnerabilities are a common concern for all users of Istio, so all users could profit from Istio images with a smaller attack surface.

We already had a look into the different images, and saw that some are already built `FROM scratch`:
* [Mixer](https://github.com/istio/istio/blob/784b031ebdde74834fa4eb45aa95370ee9fbb8ab/mixer/docker/Dockerfile.mixer#L1)
* [Sidecar Injector](https://github.com/istio/istio/blob/784b031ebdde74834fa4eb45aa95370ee9fbb8ab/pilot/docker/Dockerfile.sidecar_injector#L1)
* [Citadel](https://github.com/istio/istio/blob/784b031ebdde74834fa4eb45aa95370ee9fbb8ab/security/docker/Dockerfile.citadel#L1)

The images that use a [distribution as a base image](https://github.com/istio/istio/blob/0e9cecab053aab744a7c3a731aacb07fd794d5f9/tools/istio-docker.mk#L275-L276) are:
* [Galley](https://github.com/istio/istio/blob/784b031ebdde74834fa4eb45aa95370ee9fbb8ab/galley/docker/Dockerfile.galley#L1)
* [Pilot](https://github.com/istio/istio/blob/784b031ebdde74834fa4eb45aa95370ee9fbb8ab/pilot/docker/Dockerfile.pilot#L1)
* [Envoy](https://github.com/istio/istio/blob/784b031ebdde74834fa4eb45aa95370ee9fbb8ab/pilot/docker/Dockerfile.proxyv2#L1)
* [Proxy Init](https://github.com/istio/istio/blob/784b031ebdde74834fa4eb45aa95370ee9fbb8ab/pilot/docker/Dockerfile.proxy_init#L1)

We would like to provide Dockerfiles for the latter images that create a bare image and add their creation to the Makefiles so that they are published as additional images by the Istio build.

**Comment by kyessenov** [comment]
All components except Envoy _could be_ built from scratch. I think the only reason they are not all yet, was to add debugging tools (e.g. curl).

**Comment by esnible** [comment]
@kyessenov When we had different images for debug vs non-debug istio-proxy it was a pain.

One that that might work would be to have a `sleep`-ing debug container with tools like curl and tcpdump.  The istio-sidecar-injector could be configured to inject, or not inject, this non-functional debug container.

The other thing I do interactively for debugging is to use bash, ls, cat, and ps on the sidecar.  It might be possible to just add a few tools rather than a full distro.

**Comment by kyessenov** [comment]
@esnible All of the standard UNIX tools carry security risks. It's a matter whether we are willing to take on adding extra stuff to containers for the sake of usability.

**Comment by loewenstein** [comment]
> All components except Envoy _could be_ built from scratch. I think the only reason they are not all yet, was to add debugging tools (e.g. curl).

I'd say even Envoy _could be_ build from scratch. As a PoC I ran `objdump` and `ldd` to figure out the linked libraries.
```bash
$ objdump -s -j .interp /usr/local/bin/envoy

/usr/local/bin/envoy:     file format elf64-x86-64

Contents of section .interp:
 2002a8 2f6c6962 36342f6c 642d6c69 6e75782d  /lib64/ld-linux-
 2002b8 7838362d 36342e73 6f2e3200           x86-64.so.2.

$ ldd /usr/local/bin/envoy
	linux-vdso.so.1 =>  (0x00007fff4deaa000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007fd962fa9000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fd962da5000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fd962b88000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fd96287f000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fd9624b5000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fd9631b1000)
```

Given that we could have a two-stage Dockerfile that copies only the necessary bits to a `FROM scratch` runtime image.

```Dockerfile
FROM envoyproxy/envoy AS staging

RUN mkdir /tmp/tmp

FROM scratch

COPY --from=staging /tmp/tmp /tmp

# copy loader (taken from objdump -s -j .interp /usr/local/bin/envoy)
COPY --from=staging /lib64/ld-linux-x86-64.so.2 /lib64/ld-linux-x86-64.so.2

# copy dependent libs (taken from ldd /usr/local/bin/envoy)
COPY --from=staging /lib/x86_64-linux-gnu/librt.so.1 /lib/x86_64-linux-gnu/librt.so.1
COPY --from=staging /lib/x86_64-linux-gnu/libdl.so.2 /lib/x86_64-linux-gnu/libdl.so.2
COPY --from=staging /lib/x86_64-linux-gnu/libpthread.so.0 /lib/x86_64-linux-gnu/libpthread.so.0
COPY --from=staging /lib/x86_64-linux-gnu/libm.so.6 /lib/x86_64-linux-gnu/libm.so.6
COPY --from=staging /lib/x86_64-linux-gnu/libc.so.6 /lib/x86_64-linux-gnu/libc.so.6

COPY --from=staging /usr/local/bin/envoy /usr/local/bin/envoy

COPY --from=staging /etc/envoy/envoy.yaml /etc/envoy/envoy.yaml

CMD ["envoy", "-c", "/etc/envoy/envoy.yaml"]
```

The resulting container at least starts successfully.
```bash
$ docker run -ti envoy:minimal
[2019-02-28 10:41:56.806][1][info][main] [source/server/server.cc:202] initializing epoch 0 (hot restart version=10.200.16384.127.options=capacity=16384, num_slots=8209 hash=228984379728933363 size=2654312)
[2019-02-28 10:41:56.806][1][info][main] [source/server/server.cc:204] statically linked extensions:
[2019-02-28 10:41:56.806][1][info][main] [source/server/server.cc:206]   access_loggers: envoy.file_access_log,envoy.http_grpc_access_log
[2019-02-28 10:41:56.806][1][info][main] [source/server/server.cc:209]   filters.http: envoy.buffer,envoy.cors,envoy.ext_authz,envoy.fault,envoy.filters.http.grpc_http1_reverse_bridge,envoy.filters.http.header_to_metadata,envoy.filters.http.jwt_authn,envoy.filters.http.rbac,envoy.filters.http.tap,envoy.grpc_http1_bridge,envoy.grpc_json_transcoder,envoy.grpc_web,envoy.gzip,envoy.health_check,envoy.http_dynamo_filter,envoy.ip_tagging,envoy.lua,envoy.rate_limit,envoy.router,envoy.squash
[2019-02-28 10:41:56.806][1][info][main] [source/server/server.cc:212]   filters.listener: envoy.listener.original_dst,envoy.listener.original_src,envoy.listener.proxy_protocol,envoy.listener.tls_inspector
[2019-02-28 10:41:56.806][1][info][main] [source/server/server.cc:215]   filters.network: envoy.client_ssl_auth,envoy.echo,envoy.ext_authz,envoy.filters.network.dubbo_proxy,envoy.filters.network.mysql_proxy,envoy.filters.network.rbac,envoy.filters.network.sni_cluster,envoy.filters.network.thrift_proxy,envoy.http_connection_manager,envoy.mongo_proxy,envoy.ratelimit,envoy.redis_proxy,envoy.tcp_proxy
[2019-02-28 10:41:56.806][1][info][main] [source/server/server.cc:217]   stat_sinks: envoy.dog_statsd,envoy.metrics_service,envoy.stat_sinks.hystrix,envoy.statsd
[2019-02-28 10:41:56.806][1][info][main] [source/server/server.cc:219]   tracers: envoy.dynamic.ot,envoy.lightstep,envoy.tracers.datadog,envoy.zipkin
[2019-02-28 10:41:56.806][1][info][main] [source/server/server.cc:222]   transport_sockets.downstream: envoy.transport_sockets.alts,envoy.transport_sockets.tap,raw_buffer,tls
[2019-02-28 10:41:56.806][1][info][main] [source/server/server.cc:225]   transport_sockets.upstream: envoy.transport_sockets.alts,envoy.transport_sockets.tap,raw_buffer,tls
[2019-02-28 10:41:56.808][1][warning][misc] [source/common/protobuf/utility.cc:156] Using deprecated option 'envoy.api.v2.listener.Filter.config' from file listener.proto. This configuration will be removed from Envoy soon. Please see https://github.com/envoyproxy/envoy/blob/master/DEPRECATED.md for details.
[2019-02-28 10:41:56.808][1][warning][misc] [source/common/protobuf/utility.cc:156] Using deprecated option 'envoy.api.v2.Cluster.hosts' from file cds.proto. This configuration will be removed from Envoy soon. Please see https://github.com/envoyproxy/envoy/blob/master/DEPRECATED.md for details.
[2019-02-28 10:41:56.810][1][info][main] [source/server/server.cc:272] admin address: 127.0.0.1:9901
[2019-02-28 10:41:56.811][1][info][config] [source/server/configuration_impl.cc:50] loading 0 static secret(s)
[2019-02-28 10:41:56.811][1][info][config] [source/server/configuration_impl.cc:56] loading 1 cluster(s)
[2019-02-28 10:41:56.812][1][info][config] [source/server/configuration_impl.cc:65] loading 1 listener(s)
[2019-02-28 10:41:56.815][1][info][config] [source/server/configuration_impl.cc:90] loading tracing configuration
[2019-02-28 10:41:56.815][1][info][config] [source/server/configuration_impl.cc:110] loading stats sink configuration
[2019-02-28 10:41:56.815][1][info][main] [source/server/server.cc:472] starting main dispatch loop
[2019-02-28 10:41:56.819][1][info][upstream] [source/common/upstream/cluster_manager_impl.cc:137] cm init: all clusters initialized
[2019-02-28 10:41:56.820][1][info][main] [source/server/server.cc:446] all clusters initialized. initializing init manager
[2019-02-28 10:41:56.820][1][info][config] [source/server/listener_manager_impl.cc:1008] all dependencies initialized. starting workers
```

@mattklein123 @PiotrSikora Do you think an approach like the above would have a chance to get upstream blessing, next to [ci/Dockerfile-envoy-alpine](https://github.com/envoyproxy/envoy/blob/master/ci/Dockerfile-envoy-alpine) and [ci/Dockerfile-envoy-image](https://github.com/envoyproxy/envoy/blob/master/ci/Dockerfile-envoy-image)?

Regarding developer convenience vs. minimal attack surface. IMHO it would be fine to provide both a `*_minimal` and a `*_debug` variant of all Istio images. I find it strange to provide _only_ the debug variant though.

**Comment by sdake** [comment]
instead of `from scratch` I would encourage folks to look at the distroless containers:

https://github.com/GoogleContainerTools/distroless

**Comment by loewenstein** [comment]
@sdake didn't know that. Thanks for the pointer. Seems like `static` or `base` would be candidates for the Istio go binaries.

I will try to base the `objdump` and `ldd` trick on those as well...

**Comment by kramerul** [comment]
We created a new envoy docker image based on ```gcr.io/distroless/cc```.
We are using the `e2e_pilot` test  with the 1.1.0 release to verify that the image is working fine.
In both cases (with and without distroless) we get exactly the same 19 failures:

```
--- FAIL: TestRouteHTTPViaEgressGateway (30.65s)
    --- FAIL: TestRouteHTTPViaEgressGateway/RouteHTTPViaEgressGateway (18.42s)
--- FAIL: TestRouteSNIViaEgressGateway (54.35s)
    --- FAIL: TestRouteSNIViaEgressGateway/RouteSNIViaEgressGateway (19.40s)
    --- FAIL: TestRouteSNIViaEgressGateway/RouteSNIViaEgressGateway#01 (22.30s)
--- FAIL: TestGateway_HTTPIngress (30.51s)
    --- FAIL: TestGateway_HTTPIngress/HTTPIngressGateway (18.93s)
--- FAIL: TestGateway_HTTPSIngress (33.25s)
    --- FAIL: TestGateway_HTTPSIngress/HTTPSIngressGateway (21.69s)
--- FAIL: TestGateway_TCPIngress (29.69s)
    --- FAIL: TestGateway_TCPIngress/TCPIngressGateway (18.11s)
--- FAIL: TestCoreDumpGenerated (4.41s)
--- FAIL: TestRoutes (62.41s)
    --- FAIL: TestRoutes/v1alpha3 (57.40s)
--- FAIL: TestSidecarScope (86.92s)
    --- FAIL: TestSidecarScope/v1alpha3 (28.27s)
        --- FAIL: TestSidecarScope/v1alpha3/ns1:_http://bookinfo.com:9999_reachable_from_primary_cluster (8.34s)
        --- FAIL: TestSidecarScope/v1alpha3/ns2:_service.tcp.com:8888_reachable_from_primary_cluster (9.26s)
        --- FAIL: TestSidecarScope/v1alpha3/ns1:_bookinfo.com:9999_reachable_via_egress_TCP_listener_7.7.7.7:23145_from_primary_cluster (8.94s)
```

Therefore, we would assume that the distroless images works fine.

@costinm, @kyessenov: Which additional tests would you propose to ensure the stability of a distroless envoy?

**Comment by sdake** [comment]
@kramerul can you do me a solid and create a PR that builds envoy's container using distroless cc so we can see if this technique is functional in the gate?  IOW change your PR or create a new one that changes this:

<img width="397" alt="Screen Shot 2019-03-25 at 2 03 43 AM" src="https://user-images.githubusercontent.com/755849/54907264-4ae99000-4ea2-11e9-8803-d9fe0523ebf4.png">

to
`/pilot/docker/Dockefile.proxyv2`

The purpose is to determine how well Google's distroless container base works with proxyv2 in our gates.

Cheers
-steve

**Comment by kramerul** [comment]
@phil9909 will provide a new PR

**Comment by c0d1ngm0nk3y** [comment]
We changed all images to see if the tests are OK. Except minor issues the required tests are fine. The not required are sometimes hard to judge :(

PRs:

#12746 (scratch images)
#12741 (proxyv2)
#12788 (ubuntu based)
#12869 (proxy_init)

**Comment by sdake** [comment]
@c0d1ngm0nk3y - curious if you have tried out any applications such as bookinfo visually.  If so, could you post a screenshot?

Cheers
-steve

**Comment by c0d1ngm0nk3y** [comment]
> @c0d1ngm0nk3y - curious if you have tried out any applications such as bookinfo visually. If so, could you post a screenshot?
> 

@sdake Do I sense a lack of trust in the tests? :) It didn't came to my mind so far - but I will do so tomorrow.

**Comment by sdake** [comment]
I placed a temporary hold on the distroless work so we can discuss goals and use-cases in the env workgroup meeting on April 4th.

Cheers
-steve

**Comment by sdake** [comment]
> > @c0d1ngm0nk3y - curious if you have tried out any applications such as bookinfo visually. If so, could you post a screenshot?
> 
> @sdake Do I sense a lack of trust in the tests? :) It didn't came to my mind so far - but I will do so tomorrow.

I do have confidence in our test cases.  Our test cases do not test GUI interaction.  We need expansion of the test cases to test bookinfo GUI interactions, but until that magical work comes, a manual test case is needed.

Cheers
-steve

**Comment by sdake** [comment]
Currently an outstanding request: https://github.com/istio/istio/pull/12741#issuecomment-476702247

TLDR libnettle is needed by an operator.  We should audit (via at minimum the mailing list) what else is needed for LUA filters for envoy.

Cheers
-steve

**Comment by sdake** [comment]
The test case for generating core dumps no longer functions with this set of PRs.  https://github.com/istio/istio/pull/12741#issuecomment-476765510

This raises the question of how one would get coredumps rom the container images in general or how they would be used.

CHeers
-steve

**Comment by sdake** [comment]
@utka and @istio/test-infra-hackers I think @costinm and I are in agreement - to be discussed at the April 4th Environments WG meeting on this feature for 1.2. The next question that comes from this work is how do we test and release two sets of images.

I'd invite you to the environments wg meeting so we can all get on the same page as to why this work is needed and find a path forward. The biggest impact I see is to the T&R workgroup.

Cheers
-steve

**Comment by incfly** [comment]
/cc @quanjielin @justinburke @wattli 

Citadel agent downloads a list of public CA files for TLS connection to some CA server.
https://github.com/istio/istio/blob/master/security/docker/Dockerfile.node-agent-k8s#L3

**Comment by kramerul** [comment]
@incfly these certificates are already [part of `gcr.io/distroless/static`](https://github.com/GoogleContainerTools/distroless/blob/a4fd5de337e31911aeee2ad5248284cebeb6a6f4/base/BUILD#L84)

**Comment by holgero** [comment]
We created a cluster and installed Istio 1.1 with the distroless images. And did some manual testing:
* We installed and verified the bookinfo app.
* Verified that data is displayed in Grafana.
* Installed and tested tcp-echo and websockets from istio/samples
-> everything looks fine

**Comment by holgero** [comment]
We also ran some `istioctl` commands and saw that `istioctl proxy-status` and `istioctl authn tls-check` both fail because it tries to run commands in the container.

**Comment by jkbschmid** [comment]
The problem with `istioctl` is addressed with PR #13395

**Comment by holgero** [comment]
@sdake should we set the Milestone 1.2 on the above pr #13395 ?

**Comment by kramerul** [comment]
We created a new pull request #13579, which builds distroless docker images in addition.

**Comment by sdake** [comment]
@kramerul I very strongly want to see this feature merge in the 1.2 series.  The deadline for merge is May 21st and tests are mandatory or the work will be pushed to 1.3.

I am wrapping my head around the testing requirements given all the repo reorganization planned which put this feature at risk for 1.2 delivery.

@costinm can you provide some testing guidance?

**Comment by holgero** [comment]
My current plans to push this forward is as follows:
• We will modify the pr https://github.com/istio/istio/pull/13579 so that it doesn’t change the existing images (no change to istio-iptabels.sh in this PR), but only adds the new distroless images. It will omit building a distroless image for the init-container
• Then we will show that an installation based on the distroless images (and the existing init-container image) works. Add that as a test goal in post-submit for master or so.
• Increase the test coverage for the istio-iptables.sh script with a separate PR and provide the bugfixes for the various bugs we found in it.
• We will modify the pr https://github.com/istio/istio/pull/14071 so that it only builds the go executable that can do the same thing as the exisiting istio-iptables.sh, but won’t use it yet.
• Create a new pull request that creates an image for the init-container which is also distroless and uses the go binary instead of the shell script.
• Adapt the post-submit test from above to also use this distroless init-container image.

**Comment by kramerul** [comment]
* Distroless images are provided now: #14737
* We are running the e2e_simple test with distroless: #14944 and https://github.com/istio/test-infra/pull/1352
* We created also a distroless image for proxy_init, which uses the go binary of istio-tables.sh: #14985

After merging these pull request, @holgero 's plan should be completed.

**Comment by c0d1ngm0nk3y** [comment]
It is now part of the normal release process. So it will be available from 1.3 onwards.

```
docker pull gcr.io/istio-release/pilot:master-latest-daily-distroless
```

Hence I will close this issue.
